OptimizelySDK.Tests/DecisionServiceTest.cs

```diff
@@ -441,5 +449,281 @@ public void TestGetVariationWithBucketingId()
+            LoggerMock.Verify(l => l.Log(LogLevel.DEBUG, $"User \"user_1\" is excluded from \"Everyone Else\" rule for feature flag \"{featureFlag.Key}\"."));
+        }
+        
+        // Should return expected variation when the user is attempted to be bucketed into all targeting rules
+        // including Everyone Else rule
+        [Test]
+        public void TestGetVariationForFeatureRolloutWhenUserDoesNotQualifyForAnyTargetingRule()
+        {
+            var featureFlag = ProjectConfig.GetFeatureFlagFromKey("boolean_single_variable_feature");
+            var rolloutId = featureFlag.RolloutId;
+            var rollout = ProjectConfig.GetRolloutFromId(rolloutId);
+            var experiment0 = rollout.Experiments[0];
+            var experiment1 = rollout.Experiments[1];
+            var everyoneElseRule = rollout.Experiments[rollout.Experiments.Count - 1];
+            var expectedVariation = everyoneElseRule.Variations[0];
+
+            BucketerMock.Setup(bm => bm.Bucket(It.IsAny<ProjectConfig>(), everyoneElseRule, It.IsAny<string>(), It.IsAny<string>())).Returns(expectedVariation);
+            var decisionService = new DecisionService(BucketerMock.Object, ErrorHandlerMock.Object, ProjectConfig, null, LoggerMock.Object);
+
+            // Provide null attributes so that user does not qualify for audience.
+            var actualVariation = decisionService.GetVariationForFeatureRollout(featureFlag, "user_1", null);
+            Assert.AreEqual(expectedVariation, actualVariation);
+
+            LoggerMock.Verify(l => l.Log(LogLevel.DEBUG, $"User \"user_1\" does not meet the audience conditions to be in rollout rule \"{experiment0.Key}\"."));
+            LoggerMock.Verify(l => l.Log(LogLevel.DEBUG, $"User \"user_1\" does not meet the audience conditions to be in rollout rule \"{experiment1.Key}\"."));
+        }
+
+        #endregion // GetVariationForFeatureRollout Tests
+
+        #region GetVariationForFeature Tests
+
+        // Should return the bucketed experiment and variation 
+        [Test]
+        public void TestGetVariationForFeatureWhenTheUserIsBucketedIntoFeatureExperiment()
+        {
+            var featureFlag = ProjectConfig.GetFeatureFlagFromKey("string_single_variable_feature");
+            var expectedExperimentId = featureFlag.ExperimentIds[0];
+            var expectedExperiment = ProjectConfig.GetExperimentFromId(expectedExperimentId);
+            var expectedVariation = expectedExperiment.Variations[0];
+
+            DecisionServiceMock.Setup(ds => ds.GetVariationForFeatureExperiment(It.IsAny<FeatureFlag>(), It.IsAny<string>(), It.IsAny<UserAttributes>())).Returns(expectedVariation);
+
+            var actualVariation = DecisionServiceMock.Object.GetVariationForFeature(featureFlag, "user1", new UserAttributes());
+
+            Assert.AreEqual(expectedVariation, actualVariation);
+        }
+
+        // Should return the bucketed variation and null experiment 
```

Comment:
minor nit:
again make this comment more readable
example: When the user is not in the feature flag experiment, but is bucketed into a variation of the feature flag's rollout, return that variation. Make sure we evaluate rollouts when user doesn't satisfy experiments.

Reply:
Modified the description.